### PR TITLE
[REVIEW] Enable gather/slicing/joins with ListColumns in Python

### PR DESCRIPTION
**Pull request by shwina**
Depends on https://github.com/rapidsai/cudf/pull/5816

- [x] Adds a `cudf::column(cudf::column_view)` constructor for lists columns (`copying.cu`)
- [x] Adds tests for gather/slice/join with LIstColumns in Python

This actually started off as https://github.com/rapidsai/cudf/pull/5785 and https://github.com/rapidsai/cudf/pull/5820, but I split those off into smaller PRs because they were needed by #5723.

Fixes https://github.com/rapidsai/cudf/issues/5621

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/lists/copying/slice.cu [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+namespace lists {
+namespace detail {
+
+std::unique_ptr<cudf::column> slice(lists_column_view const& list,
+                                    size_type start,
+                                    size_type end,
+                                    cudaStream_t stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  auto sliced_table = cudf::detail::gather(table_view{{list.parent()}},
```
Comment (shwina) [inline cpp/src/lists/copying/slice.cu:15]: Just a temporary implementation

**Comment by shwina** [comment]
Just remembered I have a libcudf TODO here :)

File: python/cudf/cudf/core/column/column.py [hunk 2]
```diff
@@ -880,6 +880,11 @@ def distinct_count(self, method="sort", dropna=True):
     def astype(self, dtype, **kwargs):
         if is_categorical_dtype(dtype):
             return self.as_categorical_column(dtype, **kwargs)
+        elif is_list_dtype(dtype):
+            if not is_list_dtype(self.dtype):
+                raise ValueError(f"Cannot cast list column to {dtype}")
+            else:
+                return self
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/cudf/core/column/column.py:887]: Yeah good catch. Is that basically the only kind of casting we should can/support for list columns, i.e., casting the leaf values?

[the remark this answers is not in the document]
Comment (kkraus14) [reply python/cudf/cudf/core/column/column.py:887]: Only other semi-normal thing I can think of would be being able to do `String` <--> `List(int8)` including at various levels of nesting.

Otherwise in a bit more niche area, you could imagine someone wanting to `view` as opposed to `cast` from say `int32` --> `int64` assuming there's a divisible number of elements in each List.

File: cpp/include/cudf/lists/detail/slice.hpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/lists/lists_column_view.hpp>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+std::unique_ptr<cudf::column> slice(lists_column_view const& list,
```
Comment (shwina) [inline cpp/include/cudf/lists/detail/slice.hpp:23]: Where should tests for this go?

File: cpp/src/lists/copying/slice.cu (hunk 1 above)
Comment (shwina) [reply cpp/src/lists/copying/slice.cu]: fixed

File: cpp/include/cudf/lists/detail/gather.cuh [hunk 4]
```diff
@@ -112,6 +107,11 @@ gather_data make_gather_data(cudf::lists_column_view const& source_column,
                       return src_offsets[index];
                     });
 
+  // now that we are doing using the gather_map, we can release the underlying prev_base_offsets.
```
Comment (harrism) [inline cpp/include/cudf/lists/detail/gather.cuh:110]: Do you mean "done" rather than "doing"?

File: cpp/include/cudf/lists/detail/slice.hpp (hunk 3 above)
Comment (harrism) [reply cpp/include/cudf/lists/detail/slice.hpp:23]: tests/lists/slice_tests.cpp ???

Comment (harrism) [inline cpp/include/cudf/lists/detail/slice.hpp:23]:
```suggestion
std::unique_ptr<cudf::column> slice(lists_column_view const& lists,
```

File: cpp/src/lists/copying/slice.cu [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                                    size_type start,
+                                    size_type end,
+                                    cudaStream_t stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  size_type lists_count = lists.size();
+  if (lists_count == 0) { return cudf::empty_like(lists.parent()); }
+  if (end < 0 || end > lists_count) end = lists_count;
+  CUDF_EXPECTS(((start >= 0) && (start < end)), "Invalid start parameter value.");
+  lists_count = end - start;
```
Comment (harrism) [inline cpp/src/lists/copying/slice.cu:22]: Avoid initialize-then-modify (ITM)

```suggestion
  if (lists.is_empty() == 0) { return cudf::empty_like(lists.parent()); }
  if (end < 0 || end > lists.size()) end = lists.size();
  CUDF_EXPECTS(((start >= 0) && (start < end)), "Invalid slice range.");
  auto lists_count = end - start;
```

File: cpp/src/lists/copying/slice.cu [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+// new lists column from a subset of a lists_column_view
+std::unique_ptr<cudf::column> slice(lists_column_view const& lists,
+                                    size_type start,
+                                    size_type end,
+                                    cudaStream_t stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  size_type lists_count = lists.size();
+  if (lists_count == 0) { return cudf::empty_like(lists.parent()); }
+  if (end < 0 || end > lists_count) end = lists_count;
```
Comment (harrism) [inline cpp/src/lists/copying/slice.cu:20]: Do we need to be this flexible? Treat all negative values of `end` the same?

File: cpp/src/lists/copying/slice.cu [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  CUDF_EXPECTS(((start >= 0) && (start < end)), "Invalid start parameter value.");
+  lists_count = end - start;
+
+  auto execpol = rmm::exec_policy(stream);
```
Comment (harrism) [inline cpp/src/lists/copying/slice.cu:24]: Define just before use.

File: cpp/src/lists/copying/slice.cu [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  cudf::size_type start_offset{0};
+  cudf::size_type end_offset{0};
+  auto offsets_data = lists.offsets().data<cudf::size_type>();
+  CUDA_TRY(cudaMemcpyAsync(
+    &start_offset, offsets_data + start, sizeof(cudf::size_type), cudaMemcpyDeviceToHost, stream));
+  CUDA_TRY(cudaMemcpyAsync(
+    &end_offset, offsets_data + end, sizeof(cudf::size_type), cudaMemcpyDeviceToHost, stream));
+
+  // compute the offsets column of the result:
+  rmm::device_uvector<cudf::size_type> out_offsets(lists_count + 1, stream);
```
Comment (harrism) [inline cpp/src/lists/copying/slice.cu:40]: Why the +1? (comment).

Seems you only use lists_count+1, never just lists_count, so maybe put it into the definition of lists_count and comment there.

File: cpp/src/lists/copying/slice.cu [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+  thrust::transform(execpol->on(stream),
+                    offsets_data + start,
+                    offsets_data + end + 1,
```
Comment (harrism) [inline cpp/src/lists/copying/slice.cu:44]: Why the +1 here? // comment.

File: cpp/src/lists/copying/slice.cu [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                    out_offsets.data(),
+                    [start_offset] __device__(cudf::size_type i) { return i - start_offset; });
+  auto offsets = std::make_unique<cudf::column>(
+    cudf::data_type{cudf::type_id::INT32}, lists_count + 1, out_offsets.release());
+
+  // compute the child column of the result:
+  auto child =
```
Comment (harrism) [inline cpp/src/lists/copying/slice.cu:51]: This is a little hard to follow. Maybe a comment to explain what the two different versions of slice() being called are.

File: cpp/tests/column/column_test.cu [hunk 11]
```diff
@@ -403,4 +403,53 @@ TYPED_TEST(TypedColumnTest, ColumnViewConstructorWithMask)
   EXPECT_NE(original_view.null_mask(), copy_view.null_mask());
 }
```
Comment (harrism) [inline cpp/tests/column/column_test.cu:405]: Test the range expectations, negative start, etc.

File: cpp/include/cudf/lists/detail/slice.hpp (hunk 3 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/lists/detail/slice.hpp:23]: `cpp/tests/copying`

Comment (shwina) [reply cpp/include/cudf/lists/detail/slice.hpp:23]: Since this is really an internal function that is only ever used by the `column(column_view)` constructor, I put tests in `column_tests.cu`. I'm not sure about the name `slice` - since `slice` is also used for the API that does zero-copy slicing - but that's the name that the equivalent strings function uses. @davidwendt would you have any advice here?

File: cpp/include/cudf/lists/detail/gather.cuh (hunk 4 above)
Comment (shwina) [reply cpp/include/cudf/lists/detail/gather.cuh:110]: This change is from PR #5816 - sorry for the confusion :( cc @nvdbaranec

File: cpp/src/lists/copying/slice.cu (hunk 6 above)
Comment (shwina) [reply cpp/src/lists/copying/slice.cu]: Borrowed pattern from strings [`slice`](https://github.com/rapidsai/cudf/blob/branch-0.15/cpp/src/strings/copying/copying.cu#L31). Happy to throw here instead.

File: cpp/include/cudf/lists/detail/slice.hpp (hunk 3 above)
Comment (davidwendt) [reply cpp/include/cudf/lists/detail/slice.hpp:23]: The name of the strings function I think you are referring to in libcudf is actually called [`cudf::strings::slice_strings()`](https://docs.rapids.ai/api/libcudf/stable/group__strings__substring.html#gaa689719250048ea788b5db1549c8265f) so as to not confuse it with `cudf::slice()`

[There is a `cudf::strings::detail::slice()` that is called by `column(column_view)`](https://github.com/rapidsai/cudf/blob/ab1a1e8964f5e5b30f8aa08e81764234f24024cf/cpp/src/column/column.cu#L187) -- in the same way as `cudf::lists::detail::slice()` in [this PR](https://github.com/rapidsai/cudf/blob/ab1a1e8964f5e5b30f8aa08e81764234f24024cf/cpp/src/column/column.cu#L240).

I think adding the tests for lists for `column(column_view)` into `column_test.cu` as you have done makes sense to me.

Comment (shwina) [reply cpp/include/cudf/lists/detail/slice.hpp:23]: Right -- I guess my worry is that the `detail::slice()` functions could be confused with the public `cudf::slice()` function, and I was wondering whether we should change both `strings::detail::slice()` and `lists::detail::slice()` to something else (`copy_slice()` perhaps?)

Comment (davidwendt) [reply cpp/include/cudf/lists/detail/slice.hpp:23]: I think renaming these to `detail::copy_slice()` is a good idea. The `cudf::strings::detail::slice()` was just created before `cudf::slice()` was created/ported.

Comment (nvdbaranec) [reply cpp/include/cudf/lists/detail/slice.hpp:23]: Thoroughly agree.

File: cpp/src/lists/copying/slice.cu (hunk 7 above)
Comment (shwina) [reply cpp/src/lists/copying/slice.cu]: Done

File: cpp/src/lists/copying/slice.cu (hunk 8 above)
Comment (shwina) [reply cpp/src/lists/copying/slice.cu]: Fixed. I introduced `offsets_count` since we do actually use just `lists_count` below.

File: cpp/src/lists/copying/slice.cu (hunk 9 above)
Comment (shwina) [reply cpp/src/lists/copying/slice.cu]: Done

File: cpp/src/lists/copying/slice.cu (hunk 10 above)
Comment (shwina) [reply cpp/src/lists/copying/slice.cu]: Added a comment

File: cpp/tests/column/column_test.cu (hunk 11 above)
Comment (shwina) [reply cpp/tests/column/column_test.cu:405]: I can't test that as it's a detail of the `lists::detail::copy_slice` function (`lists::detail::slice`) function, that is not exposed in any public API.

Comment (harrism) [reply cpp/tests/column/column_test.cu:405]: I see that array_tests.[cpp, cu] have explicit calls like this `auto results      = cudf::strings::detail::copy_slice(strings_view, start, end);`.  So clearly the tests are testing `strings::detail::copy_slice`. So why can't we test `lists:detail::copy_slice`?

Comment (shwina) [reply cpp/tests/column/column_test.cu:405]: Because the `slice` function is an implementation detail, and [testing behaviour, not implementation](
https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html?m=1) is considered good practice.

File: cpp/src/lists/copying/copying.cu [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+  // Account for the offset of the view:
+  start += lists.offset();
+  end += lists.offset();
+
+  // Offsets at the beginning and end of the slice:
+  cudf::size_type start_offset{0};
+  cudf::size_type end_offset{0};
+  auto offsets_data = lists.offsets().data<cudf::size_type>();
+  CUDA_TRY(cudaMemcpyAsync(
```
Comment (nvdbaranec) [inline cpp/src/lists/copying/copying.cu:32]: Use get_value (https://github.com/rapidsai/cudf/blob/b33383c1fa644f07feb2d96aa16caf4e0445fa56/cpp/include/cudf/detail/get_value.cuh#L41) here.
```suggestion
start_offset = cudf::detail::get_value<size_type>(lists.offsets(), start, stream)
```

File: cpp/src/lists/copying/copying.cu [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    &start_offset, offsets_data + start, sizeof(cudf::size_type), cudaMemcpyDeviceToHost, stream));
+  CUDA_TRY(cudaMemcpyAsync(
```
Comment (nvdbaranec) [inline cpp/src/lists/copying/copying.cu:34]:
```suggestion
end_offset = cudf::detail::get_value<size_type>(lists.offsets(), end, stream)
```

File: cpp/src/lists/copying/copying.cu [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    cudf::data_type{cudf::type_id::INT32}, offsets_count, out_offsets.release());
+
+  // Compute the child column of the result.
+  // If the child of this lists column is itself a lists column, we call copy_slice() on it.
+  // Otherwise, it is a column of the leaf type, so we call slice() on it and copy the resulting
+  // view into a cudf::column:
+  auto child =
+    (lists.child().type() == cudf::data_type{type_id::LIST})
+      ? copy_slice(lists_column_view(lists.child()), start_offset, end_offset, stream, mr)
+      : std::make_unique<cudf::column>(
```
Comment (nvdbaranec) [inline cpp/src/lists/copying/copying.cu:57]: I haven't posted the PR yet, but I have a change in place that adds a function to lists_column_view called get_sliced_child(), which will return a column_view that has the root column offset applied.  It has some smarts in it to do the minimum amount of retrieving of offsets off the gpu.  Just worth being advised of.

Comment (shwina) [reply cpp/src/lists/copying/copying.cu:53]: I'd like to get this PR merged in ASAP if that's OK. Happy to update this to use `get_sliced_child()` in a follow-up, since I currently have a cascade of dependent PRs and I don't want to add yet another dependency

Comment (nvdbaranec) [reply cpp/src/lists/copying/copying.cu:53]: Yes, absolutely.  Just a reference for the future (I think the fate of get_sliced_child() vs. child() is probably still up for debate in that PR).

**Comment by shwina** [comment]
rerun tests

**Comment by shwina** [comment]
@harrism I'm going to retarget this to branch-0.15 which will require a force-push. Before I do that, could you please re-review/approve, as I don't want your review comments to be lost

File: cpp/src/lists/copying/copying.cu [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+          cudf::detail::slice(lists.child(), {start_offset, end_offset}, stream).front());
+
+  // Compute the null mask of the result:
+  auto null_mask = cudf::copy_bitmask(lists.null_mask(), start_offset, end_offset, stream, mr);
```
Comment (shwina) [inline cpp/src/lists/copying/copying.cu:57]: @rgsl888prabhu caught that here we should probably be doing instead:

```
  auto null_mask = cudf::copy_bitmask(lists.null_mask(), start, end, stream, mr);
```

File: cpp/tests/column/column_test.cu (hunk 11 above)
Comment (harrism) [reply cpp/tests/column/column_test.cu:405]: So by "can't", you mean "don't think I should"?

Comment (harrism) [reply cpp/tests/column/column_test.cu:405]: I guess I'm confused about why there needs to be a *behavior* that `CUDF_EXPECTS` something, but we are not allowed to test that behavior.  If we can expect a behavior, we should be able to test it.

**Comment by harrism** [comment]
> @harrism I'm going to retarget this to branch-0.15 which will require a force-push. Before I do that, could you please re-review/approve, as I don't want your review comments to be lost?

See my most recent comment about testing before I approve.

**Comment by harrism** [comment]
Seems to already be on branch-0.15. Why is a rebase needed? If you mean branch-0.16, a merge should be sufficient.

File: cpp/tests/column/column_test.cu (hunk 11 above)
Comment (shwina) [reply cpp/tests/column/column_test.cu:405]: > So by "can't", you mean "don't think I should"?

Yes -- apologies, I should have been more clear and explained my reasoning better. 

> I guess I'm confused about why there needs to be a behavior that CUDF_EXPECTS something, but we are not allowed to test that behavior. If we can expect a behavior, we should be able to test it.

TL;DR: I think we should test that behaviour if (and only if) it is exposed via a public API.

Long answer:

At the moment, we have a "private" `detail::copy_slice` function, only a part of whose functionality is exercised by a "public" `cudf::column` constructor. We have tests for that part. It's true that now we have untested code in libcudf, but it's OK because that code can not be exercised in any way using the libcudf API. If/when it is exposed publicly (e.g., in a `cudf::copy_slice`API), _then_ it becomes necessary to test for that behaviour.

In my opinion, testing that behaviour while it is private is actively harmful; consider the following scenario:

Suppose we wrote a test for the range checking the behaviour of `detail::copy_slice` today. Suppose that tomorrow, we made a public `cudf::copy_slice`. Another test for the range checking behaviour of `cudf::copy_slice` would be redundant, and understandably, could be skipped. At that point, we would have a dangerous situation where public behaviour is tested in a test for a private function, the latter of which could be dropped or changed at any time.

**Comment by shwina** [comment]
> Seems to already be on branch-0.15. Why is a rebase needed? If you mean branch-0.16, a merge should be sufficient.

Initially, this PR had base branch 0.15. Then, I changed it to 0.16. After that, I changed it again back to 0.15. As it stands, this PR wants to merge commits from 0.16 into 0.15 right now, but after I force-push my local (rebased) branch, it won't do that.

File: cpp/tests/column/column_test.cu (hunk 11 above)
Comment (harrism) [reply cpp/tests/column/column_test.cu:405]: OK, thanks for clarifying. I understand now.

**Comment by kkraus14** [comment]
@shwina can you comment here once you've rebased so we make sure that's done before we merge? 😄

**Comment by rgsl888prabhu** [comment]
It still needs a fix for mask copy which @shwina mentioned https://github.com/rapidsai/cudf/pull/5771#discussion_r468864588

**Comment by shwina** [comment]
@kkraus14 @rgsl888prabhu Force pushed, and also addressed the mask copying issue

**Comment by rgsl888prabhu** [comment]
#5816 has been merged, may be we can remove the dependency label.
